File: src/execution/slippage_guard.py
"""Pre-trade orderbook slippage estimation with adaptive sizing.

Analyzes orderbook depth before placing orders, tracks actual slippage
over a rolling window, and adjusts position sizing accordingly.
"""
from __future__ import annotations

from collections import deque
from dataclasses import dataclass

import structlog

logger = structlog.get_logger(__name__)

MAX_LEVELS_TO_WALK = 3
SPREAD_HISTORY_SIZE = 10


@dataclass(frozen=True)
class SlippageCheck:
    """Immutable result of an orderbook slippage check."""

    symbol: str
    side: str
    mid_price: float
    expected_fill_price: float
    expected_slippage_pct: float
    spread_pct: float
    can_trade: bool
    reason: str


def _walk_book_side(
    levels: list[list[float]], quantity: float
) -> tuple[float, float]:
    """Walk top levels and return (weighted_fill_price, total_available).

    Returns weighted average price consuming up to MAX_LEVELS_TO_WALK levels.
    If quantity exceeds available liquidity, returns best-effort fill price.
    """
    cost = 0.0
    filled = 0.0
    for price, qty in levels[:MAX_LEVELS_TO_WALK]:
        take = min(quantity - filled, qty)
        cost += price * take
        filled += take
        if filled >= quantity:
            break
    if filled == 0:
        return 0.0, 0.0
    return cost / filled, filled


def _calc_slippage_pct(fill_price: float, mid_price: float, side: str) -> float:
    """Calculate slippage percentage (always positive).

    Buys: fill above mid → (fill - mid) / mid * 100
    Sells: fill below mid → (mid - fill) / mid * 100
    """
    if mid_price <= 0:
        return 0.0
    if side == "buy":
        return (fill_price - mid_price) / mid_price * 100
    return (mid_price - fill_price) / mid_price * 100


class SlippageGuard:
    """Pre-trade slippage estimation with adaptive sizing."""

    def __init__(
        self,
        exchange_client,
        max_slippage_pct: float = 0.10,
        max_spread_pct: float = 0.15,
        rolling_window: int = 20,
        blacklist_threshold: int = 3,
    ) -> None:
        self._client = exchange_client
        self.max_slippage_pct = max_slippage_pct
        self.max_spread_pct = max_spread_pct
        self.rolling_window = rolling_window
        self.blacklist_threshold = blacklist_threshold
        self._spread_history: dict[str, deque[float]] = {}
        self._slippage_history: dict[str, deque[float]] = {}
        self._log = logger.bind(component="slippage_guard")

    async def check_orderbook(
        self, symbol: str, side: str, quantity: float
    ) -> SlippageCheck:
        """Analyze orderbook before placing order."""
        if side not in ("buy", "sell"):
            return self._reject(symbol, side, 0.0, 0.0, f"invalid side: {side}")

        try:
            book = await self._client.fetch_orderbook(symbol, limit=10)
        except Exception as exc:
            self._log.warning("orderbook_fetch_failed", symbol=symbol, error=str(exc))
            return self._reject(symbol, side, 0.0, 0.0, f"orderbook fetch failed: {exc}")

        bids = book.get("bids", [])
        asks = book.get("asks", [])
        if not bids or not asks:
            return self._reject(symbol, side, 0.0, 0.0, "empty orderbook")

        best_bid, best_ask = bids[0][0], asks[0][0]
        mid_price = (best_bid + best_ask) / 2
        spread_pct = (best_ask - best_bid) / mid_price * 100

        self._record_spread(symbol, spread_pct)

        if spread_pct > self.max_spread_pct:
            reason = f"spread {spread_pct:.4f}% > max {self.max_spread_pct}%"
            return SlippageCheck(
                symbol, side, mid_price, 0.0, 0.0, spread_pct, False, reason
            )

        levels = asks if side == "buy" else bids
        fill_price, available = _walk_book_side(levels, quantity)
        if available < quantity:
            fill_price, available = _walk_book_side(levels, available)

        expected_slippage = _calc_slippage_pct(fill_price, mid_price, side)

        if expected_slippage > self.max_slippage_pct:
            reason = f"expected slippage {expected_slippage:.4f}% > max {self.max_slippage_pct}%"
            return SlippageCheck(
                symbol, side, mid_price, fill_price, expected_slippage,
                spread_pct, False, reason,
            )

        if available < quantity:
            reason = f"insufficient liquidity: {available:.6f} < {quantity:.6f}"
            return SlippageCheck(
                symbol, side, mid_price, fill_price, expected_slippage,
                spread_pct, False, reason,
            )

        return SlippageCheck(
            symbol, side, mid_price, fill_price, expected_slippage,
            spread_pct, True, "",
        )

    def record_actual_slippage(
        self, symbol: str, signal_price: float, fill_price: float
    ) -> float:
        """Record actual slippage after fill. Returns slippage_pct."""
        if signal_price <= 0:
            return 0.0
        slippage_pct = abs(fill_price - signal_price) / signal_price * 100
        history = self._slippage_history.setdefault(
            symbol, deque(maxlen=self.rolling_window)
        )
        history.append(slippage_pct)
        if slippage_pct > self.max_slippage_pct:
            self._log.warning(
                "high_actual_slippage",
                symbol=symbol,
                slippage_pct=f"{slippage_pct:.4f}",
                signal_price=signal_price,
                fill_price=fill_price,
            )
        return slippage_pct

    def get_rolling_avg_slippage(self, symbol: str) -> float:
        """Average slippage over last N trades. 0.0 if no data."""
        history = self._slippage_history.get(symbol)
        if not history:
            return 0.0
        return sum(history) / len(history)

    def get_size_multiplier(self, symbol: str) -> float:
        """Position size multiplier based on rolling slippage."""
        avg = self.get_rolling_avg_slippage(symbol)
        if avg <= 0.05:
            return 1.0
        if avg > 0.08:
            return max(0.25, 0.08 / avg)
        return 1.0

    def is_blacklisted(self, symbol: str) -> bool:
        """True if last N spread checks all exceeded max_spread_pct."""
        history = self._spread_history.get(symbol)
        if not history or len(history) < self.blacklist_threshold:
            return False
        recent = list(history)[-self.blacklist_threshold:]
        if all(s > self.max_spread_pct for s in recent):
            self._log.warning("symbol_blacklisted", symbol=symbol)
            return True
        return False

    # ── private helpers ────────────────────────────────────────

    def _reject(
        self, symbol: str, side: str, mid: float, spread: float, reason: str
    ) -> SlippageCheck:
        return SlippageCheck(symbol, side, mid, 0.0, 0.0, spread, False, reason)

    def _record_spread(self, symbol: str, spread_pct: float) -> None:
        history = self._spread_history.setdefault(
            symbol, deque(maxlen=SPREAD_HISTORY_SIZE)
        )
        history.append(spread_pct)

File: pypsik-live/src/exchange/slippage_guard.rs
/// Pre-trade orderbook slippage estimation with adaptive sizing.

use std::collections::VecDeque;

use crate::exchange::client::ExchangeClient;
use tracing::warn;

const MAX_LEVELS_TO_WALK: usize = 3;
const SPREAD_HISTORY_SIZE: usize = 10;

/// Immutable result of an orderbook slippage check.
#[derive(Debug, Clone)]
pub struct SlippageCheck {
    pub symbol: String,
    pub side: String,
    pub mid_price: f64,
    pub expected_fill_price: f64,
    pub expected_slippage_pct: f64,
    pub spread_pct: f64,
    pub can_trade: bool,
    pub reason: String,
}

/// Pre-trade slippage estimation with adaptive sizing.
pub struct SlippageGuard {
    client: ExchangeClient,
    pub max_slippage_pct: f64,
    pub max_spread_pct: f64,
    rolling_window: usize,
    blacklist_threshold: usize,
    spread_history: std::collections::HashMap<String, VecDeque<f64>>,
    slippage_history: std::collections::HashMap<String, VecDeque<f64>>,
}

impl SlippageGuard {
    pub fn new(
        client: ExchangeClient,
        max_slippage_pct: f64,
        max_spread_pct: f64,
    ) -> Self {
        Self {
            client,
            max_slippage_pct,
            max_spread_pct,
            rolling_window: 20,
            blacklist_threshold: 3,
            spread_history: std::collections::HashMap::new(),
            slippage_history: std::collections::HashMap::new(),
        }
    }

    /// Analyze orderbook before placing an order.
    pub async fn check_orderbook(
        &mut self,
        symbol: &str,
        side: &str,
        quantity: f64,
    ) -> SlippageCheck {
        if side != "buy" && side != "sell" {
            return self.reject(symbol, side, 0.0, 0.0, &format!("invalid side: {}", side));
        }

        let book = match self.client.fetch_orderbook(symbol, Some(10)).await {
            Ok(b) => b,
            Err(e) => {
                warn!(symbol = symbol, error = %e, "orderbook_fetch_failed");
                return self.reject(symbol, side, 0.0, 0.0, "orderbook fetch failed");
            }
        };

        if book.bids.is_empty() || book.asks.is_empty() {
            return self.reject(symbol, side, 0.0, 0.0, "empty orderbook");
        }

        let best_bid = book.bids[0][0];
        let best_ask = book.asks[0][0];
        let mid_price = (best_bid + best_ask) / 2.0;
        let spread_pct = if mid_price > 0.0 {
            (best_ask - best_bid) / mid_price * 100.0
        } else {
            0.0
        };

        self.record_spread(symbol, spread_pct);

        if spread_pct > self.max_spread_pct {
            return SlippageCheck {
                symbol: symbol.to_string(),
                side: side.to_string(),
                mid_price,
                expected_fill_price: 0.0,
                expected_slippage_pct: 0.0,
                spread_pct,
                can_trade: false,
                reason: format!("spread {:.4}% > max {:.4}%", spread_pct, self.max_spread_pct),
            };
        }

        let levels: &[[f64; 2]] = if side == "buy" { &book.asks } else { &book.bids };
        let (fill_price, available) = walk_book_side(levels, quantity);

        let expected_slippage = calc_slippage_pct(fill_price, mid_price, side);

        if expected_slippage > self.max_slippage_pct {
            return SlippageCheck {
                symbol: symbol.to_string(),
                side: side.to_string(),
                mid_price,
                expected_fill_price: fill_price,
                expected_slippage_pct: expected_slippage,
                spread_pct,
                can_trade: false,
                reason: format!(
                    "expected slippage {:.4}% > max {:.4}%",
                    expected_slippage, self.max_slippage_pct
                ),
            };
        }

        if available < quantity {
            return SlippageCheck {
                symbol: symbol.to_string(),
                side: side.to_string(),
                mid_price,
                expected_fill_price: fill_price,
                expected_slippage_pct: expected_slippage,
                spread_pct,
                can_trade: false,
                reason: format!(
                    "insufficient liquidity: {:.6} < {:.6}",
                    available, quantity
                ),
            };
        }

        SlippageCheck {
            symbol: symbol.to_string(),
            side: side.to_string(),
            mid_price,
            expected_fill_price: fill_price,
            expected_slippage_pct: expected_slippage,
            spread_pct,
            can_trade: true,
            reason: String::new(),
        }
    }

    /// Record actual slippage after fill.
    pub fn record_actual_slippage(
        &mut self,
        symbol: &str,
        signal_price: f64,
        fill_price: f64,
    ) -> f64 {
        if signal_price <= 0.0 {
            return 0.0;
        }
        let slippage_pct = (fill_price - signal_price).abs() / signal_price * 100.0;
        let history = self
            .slippage_history
            .entry(symbol.to_string())
            .or_insert_with(|| VecDeque::with_capacity(self.rolling_window));
        history.push_back(slippage_pct);
        if history.len() > self.rolling_window {
            history.pop_front();
        }
        slippage_pct
    }

    fn reject(
        &self,
        symbol: &str,
        side: &str,
        mid: f64,
        spread: f64,
        reason: &str,
    ) -> SlippageCheck {
        SlippageCheck {
            symbol: symbol.to_string(),
            side: side.to_string(),
            mid_price: mid,
            expected_fill_price: 0.0,
            expected_slippage_pct: 0.0,
            spread_pct: spread,
            can_trade: false,
            reason: reason.to_string(),
        }
    }

    fn record_spread(&mut self, symbol: &str, spread_pct: f64) {
        let history = self
            .spread_history
            .entry(symbol.to_string())
            .or_insert_with(|| VecDeque::with_capacity(SPREAD_HISTORY_SIZE));
        history.push_back(spread_pct);
        if history.len() > SPREAD_HISTORY_SIZE {
            history.pop_front();
        }
    }
}

/// Walk top N levels of the book and return (weighted_fill_price, total_available).
fn walk_book_side(levels: &[[f64; 2]], quantity: f64) -> (f64, f64) {
    let mut cost = 0.0;
    let mut filled = 0.0;
    for [price, qty] in levels.iter().take(MAX_LEVELS_TO_WALK) {
        let take = (quantity - filled).min(*qty);
        cost += price * take;
        filled += take;
        if filled >= quantity {
            break;
        }
    }
    if filled == 0.0 {
        (0.0, 0.0)
    } else {
        (cost / filled, filled)
    }
}

/// Calculate slippage percentage (always positive).
fn calc_slippage_pct(fill_price: f64, mid_price: f64, side: &str) -> f64 {
    if mid_price <= 0.0 {
        return 0.0;
    }
    if side == "buy" {
        (fill_price - mid_price) / mid_price * 100.0
    } else {
        (mid_price - fill_price) / mid_price * 100.0
    }
}
